lib/qbo_api/raise_http_exception.rb

```diff
@@ -1,32 +1,44 @@
 require 'faraday'
+require 'faraday/response'
 require 'nokogiri'
 # @private
-module FaradayMiddleware
+class QboApi
   # @private
-  class RaiseHttpException < Faraday::Middleware
-    def call(env)
-      @app.call(env).on_complete do |response|
-        case response.status
-        when 200
-        when 400
-          raise QboApi::BadRequest.new(error_message(response))
-        when 401
-          raise QboApi::Unauthorized.new(error_message(response))
-        when 403
-          raise QboApi::Forbidden.new(error_message(response))
-        when 404
-          raise QboApi::NotFound.new(error_message(response))
-        when 429
-          raise QboApi::TooManyRequests.new(error_message(response))
-        when 500
-          raise QboApi::InternalServerError.new(error_message(response))
-        when 502
-          raise QboApi::BadGateway.new({ error_body: response.reason_phrase })
-        when 503
-          raise QboApi::ServiceUnavailable.new(error_message(response))
-        when 504
-          raise QboApi::GatewayTimeout.new(error_message(response))
-        end
+  class RaiseHttpException < Faraday::Response::RaiseError
+    def on_complete(env)
+      case env[:status]
+      when 400
+        raise QboApi::BadRequest.new(error_message(env))
+      when 401
+        raise QboApi::Unauthorized.new(error_message(env))
+      when 403
+        raise QboApi::Forbidden.new(error_message(env))
+      when 404
+        raise QboApi::NotFound.new(error_message(env))
+      when 407
+        # mimic the behavior that we get with proxy requests with HTTPS
+        msg = %(407 "Proxy Authentication Required")
+        raise Faraday::ProxyAuthError.new(msg, response_values(env))
+      when 409
+        raise Faraday::ConflictError, response_values(env)
+      when 422
+        raise Faraday::UnprocessableEntityError, response_values(env)
+      when 429
+        raise QboApi::TooManyRequests.new(error_message(env))
+      when 500
+        raise QboApi::InternalServerError.new(error_message(env))
+      when 502
+        raise QboApi::BadGateway.new({ error_body: env.reason_phrase })
+      when 503
+        raise QboApi::ServiceUnavailable.new(error_message(env))
+      when 504
+        raise QboApi::GatewayTimeout.new(error_message(env))
+      when ClientErrorStatuses
+        raise Faraday::ClientError, response_values(env)
+      when ServerErrorStatuses
+        raise Faraday::ServerError, response_values(env)
+      when nil
+        raise Faraday::NilStatusError, response_values(env)
       end
```

Comment:
Maybe should else super?

Reply:
You wrote it, you tell me :) 

Glad to change it though

Reply:
Not sure how I feel about supering into that case statement though. There are a couple places where we call `Faraday::` exceptions here, actually

Reply:
This looks good that way it is.

Reply:
It could be a hash, too

Something like:

```ruby
ERROR_STATUSES = {
  400 => QboApi::BadRequest,
  404 => QboApi::NotFound,
}

error_class = ERROR_STATUSES.fetch(env[:status]) { raise Faraday::NilStatusError }
raise error_class.new(error_message(env))
```

Looks like some of them have different variations of error messages, though, but that could be remedied.

Anyway, not a big deal. Just throwing an idea out there.